File: 36. Valid Sudoku/36. Valid Sudoku.js
/**
 * @param {character[][]} board
 * @return {boolean}
 */
var isValidSudoku = function (board) {
  for (let i = 0; i < 9; i++) {
    const seen = new Set();
    for (let j = 0; j < 9; j++) {
      const cell = board[i][j];
      if (cell !== ".") {
        if (seen.has(cell)) {
          return false;
        }
        seen.add(cell);
      }
    }
  }

  for (let j = 0; j < 9; j++) {
    const seen = new Set();
    for (let i = 0; i < 9; i++) {
      const cell = board[i][j];
      if (cell !== ".") {
        if (seen.has(cell)) {
          return false;
        }
        seen.add(cell);
      }
    }
  }

  const gridStarts = [
    [0, 0],
    [0, 3],
    [0, 6],
    [3, 0],
    [3, 3],
    [3, 6],
    [6, 0],
    [6, 3],
    [6, 6],
  ];

  for (const [startX, startY] of gridStarts) {
    const seen = new Set();
    for (let i = startX; i < startX + 3; i++) {
      for (let j = startY; j < startY + 3; j++) {
        const cell = board[i][j];
        if (cell !== ".") {
          if (seen.has(cell)) {
            return false;
          }
          seen.add(cell);
        }
      }
    }
  }

  return true;
};

File: 36. Valid Sudoku/36. Valid Sudoku.java
import java.util.HashSet;
import java.util.Set;

class Solution {

  public boolean isValidSudoku(char[][] board) {
    for (int i = 0; i < 9; i++) {
      Set<Character> seen = new HashSet<>();
      for (int j = 0; j < 9; j++) {
        char cell = board[i][j];
        if (cell != '.') {
          if (seen.contains(cell)) {
            return false;
          }
          seen.add(cell);
        }
      }
    }

    for (int j = 0; j < 9; j++) {
      Set<Character> seen = new HashSet<>();
      for (int i = 0; i < 9; i++) {
        char cell = board[i][j];
        if (cell != '.') {
          if (seen.contains(cell)) {
            return false;
          }
          seen.add(cell);
        }
      }
    }

    int[][] gridStarts = {
      { 0, 0 },
      { 0, 3 },
      { 0, 6 },
      { 3, 0 },
      { 3, 3 },
      { 3, 6 },
      { 6, 0 },
      { 6, 3 },
      { 6, 6 },
    };

    for (int[] gridStart : gridStarts) {
      int startX = gridStart[0];
      int startY = gridStart[1];
      Set<Character> seen = new HashSet<>();
      for (int i = startX; i < startX + 3; i++) {
        for (int j = startY; j < startY + 3; j++) {
          char cell = board[i][j];
          if (cell != '.') {
            if (seen.contains(cell)) {
              return false;
            }
            seen.add(cell);
          }
        }
      }
    }

    return true;
  }
}
